src/Extension/FluentSiteTreeExtension.php

```diff
@@ -219,6 +224,50 @@ protected function addLocaleStatusMessage(FieldList $fields)
         );
     }
 
+    /**
+     * Add the locale's URLSegment to the URL prefix for a page's URL segment field
+     *
+     * @param FieldList $fields
+     * @return $this
+     */
+    protected function addLocalePrefixToUrlSegment(FieldList $fields)
+    {
+        // Ensure the field is available in the list
+        $segmentField = $fields->fieldByName('Root.Main.URLSegment');
+        if (!$segmentField || !($segmentField instanceof SiteTreeURLSegmentField)) {
+            return $this;
+        }
+
+        /** @var Locale $currentLocale */
+        $currentLocale = Locale::getCurrentLocale();
+
+        /** @var Domain|null $domain */
+        $domain = $currentLocale->Domain() && $currentLocale->Domain()->exists() ? $currentLocale->Domain() : null;
+        $localeUrlSegment = $currentLocale->getURLSegment();
+
+        $baseUrl = ($domain) ? $domain->Link() : Director::absoluteBaseURL();
+
+        // Add any hierarchy segments from parent pages (see SiteTree::getCMSFields)
+        $parentSegment = SiteTree::config()->get('nested_urls') && $this->owner->ParentID
+            ? $this->owner->Parent()->RelativeLink(true)
+            : null;
+
+        // $parentSegment could contain the locale's URL segment now. If we're in domain mode then remove it
+        if ($domain && substr($parentSegment, 0, strlen($localeUrlSegment)) === $localeUrlSegment) {
+            $parentSegment = ltrim(substr($parentSegment, strlen($localeUrlSegment)), '/');
+        }
+
+        // Join the base URL component (domain, or absolute URL with locale's URL segment) to any parent page segemnts
+        $baseUrl = Controller::join_links($baseUrl, $localeUrlSegment, $parentSegment);
```

Comment:
We should join only either $baseURL and $localeURLSegment (if there is no parent), or $baseUrL and $parentSegment (if there is a parent)

Reply:
See https://github.com/tractorcow/silverstripe-fluent/issues/367#issuecomment-368662704